Lvgl simulation run PineTime screens on your PC

NeroBurner opened the pull request:
Add simulator to have pinetime Screens on PC for development.

The first Screen can be run locally, using your mouse to play Paddle.h!

![2021-10-13_lv_sim_paddle](https://user-images.githubusercontent.com/9076163/137207968-5c706105-051d-4aa7-bf6c-b4adf048ea40.png)

based on lvgl 8 port #734 

lv_sim based on: https://github.com/lvgl/lv_sim_eclipse_sdl

---
2021-11-03 updated screens with PineTime fonts

![Screenshot_20211103_214140](https://user-images.githubusercontent.com/9076163/140189966-c2e9c31f-2f23-47a0-97ae-9df89346a860.png) ![Screenshot_20211103_214323](https://user-images.githubusercontent.com/9076163/140189979-2b4ebdc7-e821-4de6-9a47-22062491108a.png) ![Screenshot_20211031_100600](https://user-images.githubusercontent.com/9076163/139575553-6d5e57ce-85a1-4f97-bb33-55d873f79265.png) 

---
2021-11-07 new Screen Metronome and the first simulated external: MotorController (shown by red dot)

![Metronome.h](https://user-images.githubusercontent.com/9076163/140663709-1e574e19-ea6c-4508-becc-bf47f5e191dc.png)

---
2021-11-15 new Screen `BatteryInfo.h` with a very low Battery, according to the `BatteryController.h` stub

![BatteryInfo.h](https://user-images.githubusercontent.com/9076163/141847871-6c7523c1-df3d-4ffb-99c0-bfb185a86e38.png)

---
2021-11-28 rebased on `develop` branch getting the new `v1.7.1` version of `FlashLight.h`

![flashlight_light](https://user-images.githubusercontent.com/9076163/143786888-d3b30307-b4ae-4df5-864b-f0f3fb1e620b.png) ![flashlight_dark](https://user-images.githubusercontent.com/9076163/143786897-13404b98-08fa-4052-851a-aed5f3449ece.png)

---
2021-11-30 All three PineTime Watch Faces are working now: `WatchFaceDigital.h`, `WatchFaceAnalog.h` and `PineTimeStyle.h`

![WatchFaceDigital](https://user-images.githubusercontent.com/9076163/143956625-c90369c4-d578-46a1-aea2-c7fdedf28cb6.png) ![WatchFaceAnalog](https://user-images.githubusercontent.com/9076163/143956202-5b577e31-9510-4270-9fcb-f74a7e840774.png) ![PineTimeStyle](https://user-images.githubusercontent.com/9076163/143956086-b1ae8a8c-17b0-4ab2-970a-4d9f76385a87.png)

---
2021-12-02 Can trigger a Notification, got the system time, and the first Settings-Screen `QuickSettings.h`

![20211202_WatchFaceDigital](https://user-images.githubusercontent.com/9076163/144509634-05aa0511-ae54-4d3c-adbe-4eed0b987fa5.png) ![20211202_WatchFaceAnalog](https://user-images.githubusercontent.com/9076163/144509660-d20c2339-92bf-44d2-b37f-96a11af44c76.png) ![20211202_PineTimeStyle](https://user-images.githubusercontent.com/9076163/144509701-dbdb5ded-2fc3-4301-9464-aab87b1f38cc.png) ![2021-12-02_Quicksettings](https://user-images.githubusercontent.com/9076163/144509913-39878aae-553b-4d0b-8623-caeb30867e78.png)

---
2021-12-03 Fix "score" alignment in `Twos.h` by upgrading lvgl submodule to the current master (`v8.1.0-64-g628299676`) 

![2021-12-03 Twos fixed align and with lvgl-8.1-master](https://user-images.githubusercontent.com/9076163/144675834-72b08442-8bf2-4c09-8c14-e1c93ae9fda2.png)
---

2022-01-16 Ported the simulator to use lvgl7 like current PineTime firmware! much less porting and maintenance burden! :tada: And `Music.h` has shown up!


![WatchfaceDigital](https://user-images.githubusercontent.com/9076163/149667267-f77b22af-8994-4b3b-9e34-e3ddfa6b9114.png) ![WatchfaceAnalog](https://user-images.githubusercontent.com/9076163/149667287-2db38d32-2fe3-4b71-9c1c-19bae0f1aa64.png) ![WatchfacePineStyle](https://user-images.githubusercontent.com/9076163/149667310-f1c83a7e-8e29-42e5-8430-9ee54aaf0ca1.png)

![Paddle](https://user-images.githubusercontent.com/9076163/149667324-e73d9f02-a858-4c1a-8f55-55ab50e2ffa1.png) ![Twos](https://user-images.githubusercontent.com/9076163/149667373-a68f9752-5449-4135-b6e9-84c7bd19f451.png) ![Metronome](https://user-images.githubusercontent.com/9076163/149667380-3d7d0288-f700-4e4a-bf7b-fb644bc8e997.png)

![Meter](https://user-images.githubusercontent.com/9076163/149667657-a2b4c2a5-82b1-49f8-a519-44142e08e2dc.png) ![Battery](https://user-images.githubusercontent.com/9076163/149667675-ed357c83-d102-44fd-b0e1-3b4b217c4acf.png)


![Torch Black](https://user-images.githubusercontent.com/9076163/149667430-0329e9f8-5bfe-4d06-b72a-b895d995b25e.png) ![Torch White](https://user-images.githubusercontent.com/9076163/149667446-8415aa24-940c-415d-bfaf-c1de9ab09a2c.png) ![QuickSettings](https://user-images.githubusercontent.com/9076163/149667462-46bce030-379a-4c12-9384-59170729b1cd.png)

![Music](https://user-images.githubusercontent.com/9076163/149672717-1254ebb6-33de-4b29-8bb3-27d45aebde67.png) ![InfiniPaint](https://user-images.githubusercontent.com/9076163/149667594-b91599de-12af-4ee3-89bc-8193024a143b.png)

---

2022-01-21: `SystemInfo.h`'s first Screen, others available, but no swipe yet

![SystemInfo Screen 1](https://user-images.githubusercontent.com/9076163/150436250-aabeaad1-4d77-49bf-bdf1-ad2c7d0afb0e.png)

---

2022-01-25: Swiping now works and most Screens are implemented

![2022-01-25_lv_sim_swiping](https://user-images.githubusercontent.com/9076163/151057090-66fa6b10-eb4f-4b62-88e6-f9f307a57e40.gif)

---

2022-01-27: add `SettingSetDate.h` and `SettingSetTime.h`

![SettingsSetDate](https://user-images.githubusercontent.com/9076163/151456513-62b18b8b-c15b-4f64-9862-bf681bef081a.png)![SettingSetTime](https://user-images.githubusercontent.com/9076163/151456554-0ef527bd-677a-409e-b84d-847ba9b4cac4.png)

---

2022-01-29: add `Screen is OFF` overlay when brightness is set to `Off`. Also add `Notifications.h`

![Screen is OFF message](https://user-images.githubusercontent.com/9076163/151674959-d3f0d7e8-9415-4ec1-9ec4-65d99e5b7623.png) ![Notifications.h](https://user-images.githubusercontent.com/9076163/151680090-d3c917ee-86b7-42a9-a482-228ecefab2b8.png)

---

2022-01-30: add the last screens `FirmwareValidator.h`, `FirmwareUpdater.h` and `PassKey.h`

![FirmwareValidator](https://user-images.githubusercontent.com/9076163/151703043-2058479a-3013-4e38-89bd-48448ffd08e8.png) ![FirmWareUpdater](https://user-images.githubusercontent.com/9076163/151702304-a207572e-e5e0-45c5-8670-17293d79d7e4.png) ![PassKey](https://user-images.githubusercontent.com/9076163/151702829-4e509ed4-e348-4852-a4ab-fe451970ea52.png)

---

2022-02-02: increase and decrease steps (`s`/`S`) and heartrate (`h`/`H`)

![WatchFaceDigital with steps and heartrate](https://user-images.githubusercontent.com/9076163/152249947-cce2b291-8863-456f-a18c-d5b59db99918.png)

Quantum-cross commented:
woah this draws lvgl screens straight from the lib without hardware?

JF002 commented:
Thanks for this PR, @NeroBurner ! InfiniTime definitely needs a simulator for the UI like this one! I have already try to build one but haven't had the time to maintain it!

Now, I'm wondering if the simulator should be included in the same project/repo as InfiniTime of if we should create another project (under the InfiniTime organization) to keep them separated? What's your opinion about this?

JF002 commented:
That's a good point! However, I'm not sure it'll be practical to maintain and manage on the long run (code, history, commits, PR,... will be mixed, making the work of maintainers and reviewer more difficult).

Wouldn't it be possible to organize the code of the simulator so that it includes InfiniTime (as a submodule) in a subfolder, for example? The CI of the simulator would check that the version of InfiniTime defined in the submodule still compiles correctly with the simulator.

NeroBurner commented:
Yes, that would be no problem. The disadvantage of an extra repo is the decoupling. If a PR breaks some assumption the simulator is using, the breakage is checked after the merge, making the simulator harder to maintain (because I (or other simulator maintainer) would have to fix the build afterwards)
The greatest advantage for the simulator (when integrated) is the possibility of checking the simulator for each PR with the CI. I don't know of a way to do that with an extra simulator-repo

JF002 commented:
It might be possible to trig an action on the repo of the simulator when a change is done in the repo of infinitime.
But I agree with you, it complicates the maintenance...
Maybe someone else have an opinion on this? @Avamander @geekbozu  ?

NeroBurner commented:
Finally fixed the Twos coloring under LVGL8 and and the swiping works as expected handling swipe events issues by LVGL :tada: 

![Screenshot_20211030_214710](https://user-images.githubusercontent.com/9076163/139556660-d6c4995a-1f72-4874-9555-6d2866b90b36.png)

NeroBurner commented:
@lectrician1 The manual adding of just one screen is just for development progress to incrementally build the simulator parts. My goal is to use unmodified PineTime Screens, but for that I need to provide the hardware specific headers as stubs. And I want to implement them bit by bit, otherwise I'd need to create everything at once, which is not feasible for me

I'm not sure if it is easier to have a separate "lv_sim" project or a target in the main project. As I stub some of the hardware dependencies (maybe FreeRTOS and nRF5 in the future) and the simulator using a different compiler (desktop amd64/arm64 GCC vs gcc-arm-none-eabi-9-2020-q2-update) having the simulator target in the main project is maybe pretty hard to manage, not unsolvable, just error prone I think

NeroBurner commented:
Time to be creative! InfiniPaint on your PC :tada: 

I reworked the code to use LVGL core widget [lv_canvas](https://docs.lvgl.io/master/widgets/core/canvas.html) and listen for LVGL touch events

![Screenshot_20211031_100600](https://user-images.githubusercontent.com/9076163/139575553-6d5e57ce-85a1-4f97-bb33-55d873f79265.png)

lectrician1 commented:
Yeah, you are right that they use different compilers. And then libraries on top of that? That could be very error prone.

But... if we do get it to a state where the entire project can build and run, I _think_ we shouldn't run into any future errors. It's not like we're going to be adding many new libraries after that. The entire project will be based around the ones we will have by then. LVGL should work, whether it's running on the PineTime or desktop.

So, I guess we should keep pressing on and making sure all components of the current project are compatible. 

I understand and like your current approach of testing each of the components. Testing the entire project immediately would be an insurmountable task.

NeroBurner commented:
One step closer to the real deal! Found the fonts, and now the `lv_sim` screens finally use the InfiniTime fonts :tada: 
![Screenshot_20211103_214140](https://user-images.githubusercontent.com/9076163/140189966-c2e9c31f-2f23-47a0-97ae-9df89346a860.png) ![Screenshot_20211103_214323](https://user-images.githubusercontent.com/9076163/140189979-2b4ebdc7-e821-4de6-9a47-22062491108a.png)

NeroBurner commented:
Next Screen `Metronome.h`. But what is this second window? And it got a dot? But only sometimes? What does it do and where does it go when it isn't on the screen? Great questions!

The second window is intended to show the simulated externals of the PineTime. Currently only the `MotorController.h` is simulated. And when the motor is active the red dot is drawn. The red dot is a placeholder, but I have just SDL to work with, and I have little experience with that. Most probably the red dot will be replaced in the future, but no promises ;)

![Screenshot_20211107_223631](https://user-images.githubusercontent.com/9076163/140663709-1e574e19-ea6c-4508-becc-bf47f5e191dc.png)

NeroBurner commented:
Found a screen, which I've never seen on my PineTime: `Meter.h`

It spins right round, right round (clockwise)

![Screenshot_20211115_204226](https://user-images.githubusercontent.com/9076163/141844429-63139db8-93be-469b-a375-eeb5d7c982c3.png)

NeroBurner commented:
My `BatteryInfo.h` says, that my Battery is low. Jokes on you! I don't have a battery on my PC! :nerd_face: 

![Screenshot_20211115_211548](https://user-images.githubusercontent.com/9076163/141847871-6c7523c1-df3d-4ffb-99c0-bfb185a86e38.png)

NeroBurner commented:
Found my simulated battery and charging cable! Now I finally can charge my simulated InfiniTime!

![Screenshot_20211118_233453](https://user-images.githubusercontent.com/9076163/142507869-89210480-5522-42ab-b100-5f3f624f22b2.png)

The simulated controllers are now changing according to keyboard-events:

- r ... enable ringing
- R ... disable ringing
- m ... let motor run
- M ... stop motor
- v ... increase remaining battery by 10 %
- V ... decrease remaining battery by 10 %
- c ... battery is charging
- C ... battery is not charging

Avamander commented:
Just wondering, could this be on a separate repository but with InfiniTime as the submodule?

NeroBurner commented:
@Avamander Yes it is possible and @JF002 had the same question as well and my answer is here https://github.com/InfiniTimeOrg/InfiniTime/pull/743#issuecomment-944967882

> Yes, that would be no problem. The disadvantage of an extra repo is the decoupling. If a PR breaks some assumption the simulator is using, the breakage is checked after the merge, making the simulator harder to maintain (because I (or other simulator maintainer) would have to fix the build afterwards)
The greatest advantage for the simulator (when integrated) is the possibility of checking the simulator for each PR with the CI. I don't know of a way to do that with an extra simulator-repo

NeroBurner commented:
There shall be `FlashLight.h`!

The four white dots on the top right of the status window show the backlight status (off, low, medium, high) and can be changed with `l` and `L`. Although the backlight has no effect on the simulator other than the status dots.

![Screenshot_20211121_223702](https://user-images.githubusercontent.com/9076163/142779866-0e9aa75a-9ab1-42ad-9e1e-bb31dc53b670.png)

and with one click it is dark

![Screenshot_20211121_224454](https://user-images.githubusercontent.com/9076163/142780071-710916d2-fc0d-4937-a082-6c022888b73c.png)

NeroBurner commented:
I got the first watchface working!!! :tada: See `WatchFaceDigital.h`

![WatchFaceDigital.h](https://user-images.githubusercontent.com/9076163/142780326-a4727a24-0d4d-4fe6-8fe3-af37073c3d67.png)

NeroBurner commented:
rebased the lvgl-8.0 branch (and this `lv_simulation` branch) on the current develop branch. In the process got the update for the `FlashLight.hpp` Screen:

![image](https://user-images.githubusercontent.com/9076163/143786888-d3b30307-b4ae-4df5-864b-f0f3fb1e620b.png)
![image](https://user-images.githubusercontent.com/9076163/143786897-13404b98-08fa-4052-851a-aed5f3449ece.png)

NeroBurner commented:
Now all three WatchFaces are working. Introducing `WatchFaceAnalog.h` and `PineTimeStyle.h`

![WatchFaceAnalog](https://user-images.githubusercontent.com/9076163/143956202-5b577e31-9510-4270-9fcb-f74a7e840774.png) ![PineTimeStyle](https://user-images.githubusercontent.com/9076163/143956086-b1ae8a8c-17b0-4ab2-970a-4d9f76385a87.png)

NeroBurner commented:
Finally, no source editing and recompile to switch between screens!

I can now switch between screens using the keyboard keys from '0' to '9'. With shift+1 translating to screen 11. So in total I can handle up to 20 screens :tada:

NeroBurner commented:
much progress, much wow!

use `n` to send a message (to get the message notification symbol on the WatchFaces). Use `N` to clear the notifications again.

While at the topic of WatchFaces, they now show the PC system time! No more Midnight in the year 1970!
![20211202_WatchFaceDigital](https://user-images.githubusercontent.com/9076163/144509634-05aa0511-ae54-4d3c-adbe-4eed0b987fa5.png) ![20211202_WatchFaceAnalog](https://user-images.githubusercontent.com/9076163/144509660-d20c2339-92bf-44d2-b37f-96a11af44c76.png) ![20211202_PineTimeStyle](https://user-images.githubusercontent.com/9076163/144509701-dbdb5ded-2fc3-4301-9464-aab87b1f38cc.png)

And the first settings screen `QuickSettings.h` is here! (Although switching away from the screen isn't yet implemented)

![2021-12-02_Quicksettings](https://user-images.githubusercontent.com/9076163/144509913-39878aae-553b-4d0b-8623-caeb30867e78.png)

NeroBurner commented:
Updating the lvgl submodule to the current master (`v8.1.0-64-g628299676`) fixes the `Twos.h` alignment issues! And now it looks like with lvgl 7.0 on Infinitime 1.7.1

![2021-12-03 Twos with lvgl-8.1-master](https://user-images.githubusercontent.com/9076163/144671160-c72dc7a5-8029-4a0a-963e-c9d521d2b0ac.png)

NeroBurner commented:
but now, behold! `Twos.h` styled as with 1.7.1!

- white font for tiles with value >=8
- find and update style to set `LV_TEXT_ALIGN_CENTER`

![2021-12-03 Twos fixed](https://user-images.githubusercontent.com/9076163/144675834-72b08442-8bf2-4c09-8c14-e1c93ae9fda2.png)

NeroBurner commented:
Ported the simulator to LVGL7! Much less maintenance burden on my side and easier for others to try it out

![WatchfaceDigital](https://user-images.githubusercontent.com/9076163/149667267-f77b22af-8994-4b3b-9e34-e3ddfa6b9114.png) ![WatchfaceAnalog](https://user-images.githubusercontent.com/9076163/149667287-2db38d32-2fe3-4b71-9c1c-19bae0f1aa64.png) ![WatchfacePineStyle](https://user-images.githubusercontent.com/9076163/149667310-f1c83a7e-8e29-42e5-8430-9ee54aaf0ca1.png)

![Paddle](https://user-images.githubusercontent.com/9076163/149667324-e73d9f02-a858-4c1a-8f55-55ab50e2ffa1.png) ![Twos](https://user-images.githubusercontent.com/9076163/149667373-a68f9752-5449-4135-b6e9-84c7bd19f451.png) ![Metronome](https://user-images.githubusercontent.com/9076163/149667380-3d7d0288-f700-4e4a-bf7b-fb644bc8e997.png)

![Meter](https://user-images.githubusercontent.com/9076163/149667657-a2b4c2a5-82b1-49f8-a519-44142e08e2dc.png) ![Battery](https://user-images.githubusercontent.com/9076163/149667675-ed357c83-d102-44fd-b0e1-3b4b217c4acf.png)


![Torch Black](https://user-images.githubusercontent.com/9076163/149667430-0329e9f8-5bfe-4d06-b72a-b895d995b25e.png) ![Torch White](https://user-images.githubusercontent.com/9076163/149667446-8415aa24-940c-415d-bfaf-c1de9ab09a2c.png) ![QuickSettings](https://user-images.githubusercontent.com/9076163/149667462-46bce030-379a-4c12-9384-59170729b1cd.png) ![InfiniPaint](https://user-images.githubusercontent.com/9076163/149667594-b91599de-12af-4ee3-89bc-8193024a143b.png)

Avamander commented on lv_sim/CMakeLists.txt:
```diff
@@ -0,0 +1,146 @@
+target_link_libraries(main PRIVATE SDL2::SDL2)
+
+
+# freeRTOS headers
+#target_include_directories(main PRIVATE "../src") # include for FreeRTOSConfig.h
+#target_include_directories(main PRIVATE "../src/FreeRTOS") # include for FreeRTOSConfig.h
+#target_sources(main PRIVATE
+#  ../src/FreeRTOS/port_cmsis_systick.c
+#  ../src/FreeRTOS/port_cmsis.c
+#  ../src/FreeRTOS/port.c)
+##target_include_directories(main PRIVATE "/home/nero/repos/pinetime/FreeRTOS-LTS/FreeRTOS/FreeRTOS-Kernel/include")
```
A bunch of things like these have been commented out, if they're not required can we delete them?

There's also mentions of your home folder that should be removed.

NeroBurner replied:
yes, will remove, thanks for looking into it!

NeroBurner replied:
updated

NeroBurner commented:
`SystemInfo.h`'s first Screen is here to stay!

![SystemInfo Screen 1](https://user-images.githubusercontent.com/9076163/150436250-aabeaad1-4d77-49bf-bdf1-ad2c7d0afb0e.png)

Also much background work for the simulator like implementing FreeRTOS `timer.h` using SDL-timers and the TouchPanel-driver (including gestures) implemented polling SDL mouse. Unfortunately the gestures and mouse isn't yet hooked up with `DisplayApp.h` or `SystemTask.h`. So no actual swiping to switch screens yet, but little by little I'm getting closer

NeroBurner commented:
Swiping, Longpress and general interaction is now implemented. Most of the Screens are implemented as well (missing are FirmwareUpdate, FirmwareValidation, Notifications, SettingSetDate and SettingSetTime)

![2022-01-25_lv_sim_swiping](https://user-images.githubusercontent.com/9076163/151057090-66fa6b10-eb4f-4b62-88e6-f9f307a57e40.gif)

NeroBurner commented:
two screens added to the simulated screen list: `SettingSetDate.h` and `SettingSetTime.h`, and with reenabling some `SystemTask.cpp` code the date and time setting actually stick (as long as the simulator runs, on start the simulator time is initialized with the system-time)

![SettingsSetDate](https://user-images.githubusercontent.com/9076163/151456513-62b18b8b-c15b-4f64-9862-bf681bef081a.png)![SettingSetTime](https://user-images.githubusercontent.com/9076163/151456554-0ef527bd-677a-409e-b84d-847ba9b4cac4.png)

NeroBurner commented:
Added an overlay if the screen is turned off (brightness-level is `Off`). Would be confusing to have an enabled screen, but input is just ignored. So let the user know!

![Screen is OFF message](https://user-images.githubusercontent.com/9076163/151674959-d3f0d7e8-9415-4ec1-9ec4-65d99e5b7623.png)

NeroBurner commented:
We can now use `SystemTask.cpp` from `src` instead of the modified simulated copy :tada:

NeroBurner commented:
Finally can read my simulated Notifications!

![Notifications.h](https://user-images.githubusercontent.com/9076163/151680090-d3c917ee-86b7-42a9-a482-228ecefab2b8.png)

NeroBurner commented:
The wait for the `FirmwareUpdater.h` Screen is over!!! Although waiting for the actual (simulated) update to happen... hmmm... still waiting

![FirmWareUpdater](https://user-images.githubusercontent.com/9076163/151702304-a207572e-e5e0-45c5-8670-17293d79d7e4.png)

Accessible currently with the screen `7`

NeroBurner commented:
That must be one of the most sophisticated passkey ever existed... how do they know my super secure password! o.O

![PassKey](https://user-images.githubusercontent.com/9076163/151702829-4e509ed4-e348-4852-a4ab-fe451970ea52.png)

And with `PassKey.h`, we got all the screens in the simulator, that are listed/used in `DisplayApp.h`, AND we're using the main `DisplayApp.h` and cpp files (as well as the main `SystemTask.h`+cpp) :tada:

NeroBurner commented:
I like it when the firmware is validated, and with it the simulator is validated, isn't it? :grin: 

![FirmwareValidator](https://user-images.githubusercontent.com/9076163/151703043-2058479a-3013-4e38-89bd-48448ffd08e8.png)

FintasticMan commented on lv_sim/sim/queue.cpp:
```diff
@@ -0,0 +1,46 @@
+BaseType_t xQueueSendFromISR(QueueHandle_t &xQueue, const void * const pvItemToQueue, BaseType_t *xHigherPriorityTaskWoken)
+{
+  TickType_t xTicksToWait = 0;
+  xHigherPriorityTaskWoken = pdFALSE;
```
I think this line should be `*xHigherPriorityTaskWoken = pdFALSE;`. This builds fine as it is with gcc, but with clang this throws an error.

NeroBurner replied:
you're absolutely right! We should never reach that code because the isr-registers are stubbed for the `is_isr()` function in `DisplayApp.cpp` and `SystemTask.cpp` to always return `false`. But better fix it, especially when it breaks a compiler!

```suggestion
  *xHigherPriorityTaskWoken = pdFALSE;
```

NeroBurner commented:
rebased on current develop branch with 24 hour `SettingSetTime.h` and `Alarm.h`

![SetTime 24h](https://user-images.githubusercontent.com/9076163/152051970-26633c03-78da-46b2-8559-440b4f0a6fee.png) ![Alarm with 24 hours](https://user-images.githubusercontent.com/9076163/152051867-a8820d20-a5e3-4352-97ea-6f58c50c88b1.png)

Renamed the resulting executable to `pinesim`. I hope you like the new name, the old `main` was a bit boring

And we now have a workflow for `lv_sim`, which also provides a working `pinesim` executable for linux

NeroBurner commented:
you can now increase your simulated steps away from the immovable `0000`. Increase steps by `500` by pressing `s`, and decrease with `S` (`shift+s`)

Now after SO many steps, I bet your heartrate is high, let it be as high as you want! Increase heartrate by 10 by pressing `h` and stop the heartrate with `H` (`shift+h`)

![WatchFaceDigital with steps and heartrate](https://user-images.githubusercontent.com/9076163/152249947-cce2b291-8863-456f-a18c-d5b59db99918.png)

JF002 commented:
As I said previously : I'm astonished by the amount of work you put in this simulator : CMake, simulator code, reuse of InfiniTime code, stub of hardware depend code, everything is there! There have been previous attempt at a simulator for InfiniTime (I have started one myself), but this one is the most advanced and complete :100:  !

This simulator will be very helpful to anyone wishing to design new UIs and apps in InfiniTime, as it allows to test and debug directly on the computer, without needed to flash the PineTime for each test!

In addition, this simulator paves the way for porting InfiniTime to another platform : all the hardware dependent code have already been identified! Now, we "*just*" need to encapsulate this code correctly so we can easily run InfiniTime on other platform or MCU. One of my target being the BL60x on the PineDio Stack (future dev kit from Pine64).

One think I would like to improve before merging is the integration of the simulator in the project. For now, the simulator is a project inside another project, with no real glue between the two : you cannot build the simulator from the main CMake file. Also, the toolchain and dependencies of the simulator are very different from the one needed by InfiniTime (native vs ARM toolchain, SDL drivers for LVGL, hardware for InfiniTime,...
So I was wondering if we should keep it that way, and try to improve the integration of both projects to ensure the dependcies and toolchains are correctly managed by CMake.
Or maybe we should split both project : keep InfiniTime as it is right now, and create a whole new project for the simulator.  This new project would include InfiniTime as a git submodule, for example.

This 2nd solution has several advantages:
 - Better separation between the 2 projects, no confusion with dependencies and toolchains
 - The git history are separated
 - PR for the simulator are also separated from the ones for InfniTime, which will probably allow for faster and easier merges.
 - The simulator can specify which version of InfiniTime it pulls (via git submodule) so that a new change in InfiniTime won't break the simulator. This will allow the devs of the simulator to include those new changes when they are ready to support them in the simulator

Of course, there are also a few disadvantages
 - 2 projects to maintain instead of one
 - Keeping both project in sync will require a bit more rigore

I talked about this with @NeroBurner and we agree that splitting looks like a better idea. We will probably create the new project in the InfiniTime organization very soon !

JF002 commented:
Here are a few more info about how both project will work together:

As I said, InfiniTime repo will only contain the InfiniTime code, the code that will be flashed into you PineTime. The new repo will contain the code specific to the simulator and will include the whole InfiniTime project as a git submodule.

As I said earlier, this will make the maintenance of both projects easier. As you've probably noticed, it takes quite a long time for PRs to be reviewed and merged. That's partly because changes in InfiniTime need extensive testing and reviews to ensure good code quality and that those changes will actually run on the watch (while avoiding crashes, memory issues,...). If the simulator was included in InfiniTime, the time an efforts needed to merge a PR would increase : we'll need to check that the simulator still work with those new changes, otherwise, the simulator would be broken for everyone on `develop`.

Splitting allows to reduce the coupling between those projects: there'll be maintainers more dedicated to InfiniTime and others dedicated to the simulator. The simulator will include a specific version (commit) of InfiniTime, so that breaking changes in Infinitime won't be applied to the simulator until developers had the time to adapt to those changes. PR in the simulator will also be easier to merge as there are less constraints in code quality and testing for the simulator.

As I have mentionned earlier, the simulator helps to see what exactly needs to be done in InfiniTime to better abstract the hardware. In the future, I would like to improve those abstractions in InfiniTime so the simulator does not need to stub that many modules from InfiniTime. This will also help porting to other MCUs and writiing unit tests for both projects.

Now, how will you use those projects? If you want to develop on InfiniTime, nothing changes : clone, checkout, edit, build, test, commit, push,...
If you want to use InfiniTime with the simulator, just checkout the simulator and init/update the git submodule. It'll download the lastest version of InfiniTime that is know to work with the simulator. Then build, run and enjoy.
If you want to run the simulator against another version of InfiniTime, just checkout that version in the subfolder that contains the submodule, build and run!
If you want to create new apps for InfiniTime in the simulator, you can edit InfiniTime inside the git submodule folder (create a branch first), or point the simulator to the folder that contains your working copy of InfiniTime (via a CMake command line argument).

Of course, we will also probably create new Github Actions to build the simulator. It might also be possible to trigger an action that builds the simulator each time a new commit is pushed on InfiniTime. This way, we'll directly know if this commit breaks the simulator.

Note that the separate project does not exist yet, but @NeroBurner accepted to join the InfiniTime organization to work on and maintain the simulator! Welcome to the team, @NeroBurner !

NeroBurner commented:
Thank you for your kind words. I'm happy to be part of such an awesome project.

---

Successfully created https://github.com/InfiniTimeOrg/InfiniSim project :tada: 

the checked in submodule points to this branch because of some fixes needed to build the simulator

---

On this side the CI now downloads the InfiniSim project to compile the `infinisim` binary for all PRs in the future, to ensure that the simulator stays working in the future

NeroBurner commented:
A small overview of the related PRs important for this PR.

This PR contains the commits of other PRs fixing various things to make the Simulator work:

- https://github.com/InfiniTimeOrg/InfiniTime/pull/967
- https://github.com/InfiniTimeOrg/InfiniTime/pull/975
- https://github.com/InfiniTimeOrg/InfiniTime/pull/976
- https://github.com/InfiniTimeOrg/InfiniTime/pull/977
- https://github.com/InfiniTimeOrg/InfiniTime/pull/979
- https://github.com/InfiniTimeOrg/InfiniTime/pull/995
- https://github.com/InfiniTimeOrg/InfiniTime/pull/996
- https://github.com/InfiniTimeOrg/InfiniTime/pull/1000

Then the [lv_sim: CI: add lv_sim workflow and upload 'infinisim' executable](https://github.com/InfiniTimeOrg/InfiniTime/pull/743/commits/beb403fe363a0965c92d654fdfdfc180ac749a71) is the final commit, which is very important to help keep the InfiniSim project in sync with the main InfiniTime project

JF002 commented:
@NeroBurner I've just merged the last PR that was needed for the simulator. I think we can close this issue now that InfiniSim has [its own project](https://github.com/InfiniTimeOrg/InfiniSim) :)

NeroBurner commented:
Not yet, this branch contains the commit to compile the InfiniSim simulator on each pull request, which would be great to have to be able to keep compatibility between InfiniTime and InfiniSim. I'll cleanup and rebase this branch right now
